server/sonar-server/src/main/java/org/sonar/server/computation/source/SymbolsLineReader.java

```diff
@@ -0,0 +1,115 @@
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+package org.sonar.server.computation.source;
+
+import org.sonar.batch.protocol.output.BatchReport;
+import org.sonar.server.source.db.FileSourceDb;
+
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.List;
+import java.util.Map;
+
+import static com.google.common.collect.Lists.newArrayList;
+import static com.google.common.collect.Maps.newHashMap;
+
+public class SymbolsLineReader implements LineReader {
+
+  private final List<BatchReport.Symbols.Symbol> symbols;
+  private final Map<BatchReport.Symbols.Symbol, Integer> idsBySymbol;
+
+  public SymbolsLineReader(List<BatchReport.Symbols.Symbol> symbols) {
+    this.symbols = symbols;
+    sortSymbols();
+    this.idsBySymbol = createIdsBySymbolMap(symbols);
+  }
+
+  @Override
+  public void read(FileSourceDb.Line.Builder lineBuilder) {
+    int line = lineBuilder.getLine();
+    List<BatchReport.Symbols.Symbol> lineSymbols = findSymbolsMatchingLine(line);
+    for (BatchReport.Symbols.Symbol lineSymbol : lineSymbols) {
+      int symbolId = idsBySymbol.get(lineSymbol);
+      StringBuilder symbolString = new StringBuilder(lineBuilder.getSymbols());
+
+      appendSymbol(symbolString, lineSymbol.getDeclaration(), line, symbolId, lineBuilder.getSource());
+      for (BatchReport.Range range : lineSymbol.getReferenceList()) {
+        appendSymbol(symbolString, range, line, symbolId, lineBuilder.getSource());
+      }
+
+      if (symbolString.length() > 0) {
+        lineBuilder.setSymbols(symbolString.toString());
+      }
+    }
+  }
+
+  private void appendSymbol(StringBuilder lineSymbol, BatchReport.Range range, int line, int symbolId, String sourceLine) {
+    if (matchLine(range, line)) {
+      RangeHelper.appendRange(lineSymbol, range, line, sourceLine.length());
+      lineSymbol.append(symbolId);
+    }
+  }
+
+  private List<BatchReport.Symbols.Symbol> findSymbolsMatchingLine(int line) {
+    List<BatchReport.Symbols.Symbol> lineSymbols = newArrayList();
+    for (BatchReport.Symbols.Symbol symbol : symbols) {
+      if (matchLine(symbol.getDeclaration(), line)) {
+        lineSymbols.add(symbol);
+      } else {
+        for (BatchReport.Range range : symbol.getReferenceList()) {
+          if (matchLine(range, line)) {
+            lineSymbols.add(symbol);
+          }
+        }
+      }
+    }
+    return lineSymbols;
+  }
+
+  private static boolean matchLine(BatchReport.Range range, int line) {
+    return range.getStartLine() <= line && range.getEndLine() >= line;
+  }
+
+  private void sortSymbols() {
```

Comment:
I wanted to compare with batch way of sorting to check if it's the same ordering